node/src/Pos/Explorer/DB.hs

```diff
@@ -104,10 +133,116 @@ prepareExplorerDB = do
             addressCoinPairs = utxoToAddressCoinPairs utxo
         putGenesisBalances addressCoinPairs
         putInitFlag
+
     -- Smooth migration for CSE-228.
     unlessM utxoSumInitializedM $ do
         putCurrentUtxoSum
 
+    -- Smooth migration for CSE-236.
+    unlessM blockEpochPagesInitializedM $ do
+        convertOldEpochBlocksFormat
+  where
+    -- | If we have persisted the 0 @Epoch@ with the old format, we need to migrate it.
+    -- Since we start syncing from the start, this will detect if the user has the old
+    -- or the new format.
+    -- Returns @True@ when the old format doesn't exist.
+    blockEpochPagesInitializedM :: MonadDBRead m => m Bool
+    blockEpochPagesInitializedM = isNothing <$> dbGet GStateDB existingOldEpochBlocks
+      where
+        existingOldEpochBlocks :: ByteString
+        existingOldEpochBlocks = previousEpochBlocksPrefix $ EpochIndex minBound
+
+    -- | This is where we convert the old format of @Epoch@ which has 21600 blocks to the
+    -- new and shiny one with @Page@s added.
+    convertOldEpochBlocksFormat :: (MonadDBRead m, MonadDB m) => m ()
+    convertOldEpochBlocksFormat =
+        runConduitRes $  epochsSource
+                      .| epochPagesConduit
+                      .| epochPagesExplorerOpConduit
+                      .| persistExplorerOpSink
+      where
+        -- 'Source' corresponding to the old @Epoch@ format that contained all
+        -- 21600 @[HeaderHash]@ in a single @Epoch@ (in production).
+        epochsSource :: (MonadDBRead m) => Source (ResourceT m) (Epoch, [HeaderHash])
+        epochsSource = dbIterSource GStateDB (Proxy @EpochsIter)
+
+        -- 'Conduit' to turn the old @Epoch@ format that contained all
+        -- 21600 @[HeaderHash]@ in a single @Epoch@ to a new format that contains
+        -- @Epoch@ *and* @Page@ as a key since there are too many @[HeaderHash]@ to
+        -- load - performance reasons.
+        epochPagesConduit
+            :: (MonadDBRead m)
+            => Conduit (Epoch, [HeaderHash]) m (Map EpochPagedBlocksKey [HeaderHash])
+        epochPagesConduit = CL.map convertToPagedMap
+          where
+            convertToPagedMap
+                :: (Epoch, [HeaderHash])
+                -> Map EpochPagedBlocksKey [HeaderHash]
+            convertToPagedMap ehh = fromListWith (++) (convertToPaged ehh)
+
+            convertToPaged :: (Epoch, [HeaderHash]) -> [((Epoch, Page), [HeaderHash])]
+            convertToPaged (epoch, headerHashes) =
+                convertHHsToEpochPages <$> convertHHsToPages
```

Comment:
Please replace `<$>` with `map` here.

Reply:
Why? It's not like this is the only piece of code that has the more general version of the functor.
If it's because of readability, I can change it, but since the functions are (type) annotated, I think it's pretty readable.

Reply:
@ksaric generalization is good when you intend to use it. In this case it's unneeded. Furthermore, `map` reminds the reader of the type of its arguments instantly, unlike `<$>`. Please don't put the extra strain of keeping track of type information on the reader and use `map`.